bluesky/run_engine.py

```diff
@@ -1602,9 +1615,11 @@ def fut_fac() -> awaitable:
         the RunStart document
         """
         if self._run_is_open:
-            raise IllegalMessageSequence("A 'close_run' message was not "
-                                         "received before the 'open_run' "
-                                         "message")
+            exception = IllegalRunSequence(
```

Comment:
It is normal for user-defined exceptions to take custom arguments in their `__init__`. See [documented examples](https://docs.python.org/3/tutorial/errors.html#user-defined-exceptions). I didn't know that when I wrote the code in event-model that I presented as a model for this approach, and so I chose to monkey-patch instead, but setting things up at `__init__` time is always better if it's possible.

So this could be:

```py
IllegalRunSequence(
    "A 'close_run' message was not received before the 'open_run' "
    "message", msg, self._currently_open_runs)
```

which to my reading seems more self-contained / easy to follow. It removes `self._exception_attrs` and `_add_attrs_to_exception` entirely, and along with them the need to cross-reference anything to follow the action of the code. Each exception type can take in only those things which are useful/relevant. That is, some may want `self._currently_open_runs`; some way want `self._state`; and so on.

Reply:
I just opened a PR to fix event-model to follow this approach as well. The relevant commit is https://github.com/bluesky/event-model/pull/104/commits/bb9618ec639687bcf8e38587f5e40246972f3176.

Reply:
good call @danielballan, this is a much neater solution. I will make this change this afternoon.

Reply:
fixed